pkg/telemetrymetadata/stmt_parse.go

```diff
@@ -0,0 +1,132 @@
+package telemetrymetadata
+
+import (
+	"strings"
+
+	"github.com/AfterShip/clickhouse-sql-parser/parser"
+	"github.com/SigNoz/signoz/pkg/types/telemetrytypes"
+)
+
+// TelemetryFieldVisitor is an AST visitor for extracting telemetry fields
+type TelemetryFieldVisitor struct {
+	parser.DefaultASTVisitor
+	Fields []*telemetrytypes.TelemetryFieldKey
+}
+
+func NewTelemetryFieldVisitor() *TelemetryFieldVisitor {
+	return &TelemetryFieldVisitor{
+		Fields: make([]*telemetrytypes.TelemetryFieldKey, 0),
+	}
+}
+
+// VisitColumnDef is called when visiting a column definition
+func (v *TelemetryFieldVisitor) VisitColumnDef(expr *parser.ColumnDef) error {
+	// Check if this is a materialized column with DEFAULT expression
+	if expr.DefaultExpr == nil {
+		return nil
+	}
+
+	// Parse column name to extract context and data type
+	columnName := expr.Name.String()
+
+	// Remove backticks if present
+	columnName = strings.TrimPrefix(columnName, "`")
+	columnName = strings.TrimSuffix(columnName, "`")
+
+	// Parse the column name to extract components
+	parts := strings.Split(columnName, "_")
+	if len(parts) < 2 {
+		return nil
+	}
+
+	context := parts[0]
+	dataType := parts[1]
+
+	// Check if this is a valid telemetry column
+	var fieldContext telemetrytypes.FieldContext
+	switch context {
+	case "resource":
+		fieldContext = telemetrytypes.FieldContextResource
+	case "scope":
+		fieldContext = telemetrytypes.FieldContextScope
+	case "attribute":
+		fieldContext = telemetrytypes.FieldContextAttribute
+	default:
+		return nil // Not a telemetry column
+	}
+
+	// Check and convert data type
+	var fieldDataType telemetrytypes.FieldDataType
+	switch dataType {
+	case "string":
+		fieldDataType = telemetrytypes.FieldDataTypeString
+	case "bool":
+		fieldDataType = telemetrytypes.FieldDataTypeBool
+	case "int", "int64":
+		fieldDataType = telemetrytypes.FieldDataTypeFloat64
+	case "float", "float64":
+		fieldDataType = telemetrytypes.FieldDataTypeFloat64
+	case "number":
+		fieldDataType = telemetrytypes.FieldDataTypeFloat64
+	default:
+		return nil // Unknown data type
+	}
+
+	// Extract field name from the DEFAULT expression
+	// The DEFAULT expression should be something like: resources_string['k8s.cluster.name']
+	// We need to extract the key inside the square brackets
+	defaultExprStr := expr.DefaultExpr.String()
+
+	// Look for the pattern: map['key']
+	startIdx := strings.Index(defaultExprStr, "['")
+	endIdx := strings.Index(defaultExprStr, "']")
```

Comment:
where are we handling top level columns ?

Reply:
If by top-level columns, you mean intrinsic fields of the log record like trace_id, severity text, then this change is not about it. This is about extracting the materialized columns from them the table schema.

Reply:
Got your point. The title makes it clear.